### DISCUSSION TITLE
What is the cause of “Loading initial props cancelled”?

### DISCUSSION THREAD

**Discussion by janaroj** [vercel/next.js#21073] [Help]
In our application we have a few different users (~20 according to Plumbr) per day who get the following exception

`Error: Loading initial props cancelled`

We have determined that the error comes from the `_getData<T>` method in router.ts but haven't been able to figure out the cause nor reproduce it. We have tried adding throttling using Chrome dev tools and doing multiple actions that call router in quick succession but no dice.

What causes the error?

**Comment by willbamford** [comment]
Has anyone identified how / when this error is raised ([the source](https://github.com/vercel/next.js/blob/cd3e054f14ce38f4ff57c727a997da2a6e1d05dd/packages/next/shared/lib/router/router.ts#L2191))? Since we upgraded from Next.js v12.0.10 to v12.2.2 we started receiving a lot of these errors through our Browser error tracker (from no errors to lots). As far as we're aware this isn't impacting customers but we aren't 100% sure so not confident we are safe to filter these error logs out.

**Reply by etienne-martin** [reply]
From what I understand, it happens when navigating away from a page that is still loading. This is more common for users with slow connections. You can reproduce the error by throttling down your network speed in chrome and navigating to multiple pages that have `getInitialProps` in short succession. 

I'd say it's safe to ignore.

**Reply by guillermodlpa** [reply]
This is correct for me, it happens consistently when navigating away from a page that is still loading when that page has `getServerSideProps`. However, it seems that it wasn't OK for us to ignore because some mobile browsers are detecting when errors happen repeatedly, and show this screen to users that says _"website" has generated problems repeatedly_. This screenshot was taken after producing some of those errors in Safari, opened from a link in Slack:

![Screenshot of error screen](https://user-images.githubusercontent.com/1557348/212005027-4abe5287-2b1c-4162-8920-5290eaacd3d4.png)

We solved it by debouncing when the URL is updated, putting the `router.replace`/`router.push` inside a `useEffect` with a timeout. [See a gist with the code.](https://gist.github.com/guillermodlpa/1148b2a0409c2730f65abc49c25a49fb) The error can still happen if a connection is slow and the URL updates while there's still another page loading, but we reduced the frequency enough that we rarely see it again.

Hope this helps!

**Reply by TGalstyan1994** [reply]
You can clearInterval before router.push or replace
<img width="594" alt="image" src="https://github.com/vercel/next.js/assets/46672382/36b0a816-f81b-43e9-a57e-8d0b45578911">

**Comment by JasperAlexander** [comment]
Setting a timeout of 1500 ms works for me, but this is not ideal.

**Reply by jonataspinto** [reply]
Hi @JasperAlexander! Setting this timeout where? And why is it not ideal?

**Reply by JasperAlexander** [reply]
> Hi @JasperAlexander! Setting this timeout where? And why is it not ideal?

I got this error when trying to redirect a page visitor to another page using router.push based on an api fetch inside a useEffect hook. After using a timeout using setTimeout of 1500 ms inside the useEffect hook, this error was solved. However, it wouldn't surprise me if you still get this error with a slower internet connection. Another reason why the solution is not ideal, is because 1500 ms is quite a long time.

If you want to redirect a page visitor to another page based on an api fetch, you can better use redirect inside getServerSideProps.

**Reply by jonataspinto** [reply]
hmmm i get it, makes sense! But in my case it happens with the `<Link/>` . That's why I asked about the timeout

**Reply by loulou1994** [reply]
> hmmm i get it, makes sense! But in my case it happens with the `<Link/>` . That's why I asked about the timeout

I'm using the Link component as well. Did you figure something out to resolve the issue?

**Reply by jonataspinto** [reply]
> > hmmm i get it, makes sense! But in my case it happens with the `<Link/>` . That's why I asked about the timeout
> 
> I'm using the Link component as well. Did you figure something out to resolve the issue?

Not yet, we decided to ignore it for now

**Reply by flyingdev** [reply]
I use <Link> tag as well. what could the solution?

**Reply by imraann00** [reply]
did you guys find a solution for this

**Comment by TylerP33** [comment]
I can't seem to find a good solution for this, but this error pops up when I am passing query params using router.push.  For whatever reason, the error goes away when shallow routing is set to true.  If shallow routing is set to false, it seems there's a conflict between state variables that are being listened to in the useEffect dependency array and router.push.  The synchronization of that is causing some problems, but I am not exactly sure why - if someone wants to dive in, this may help.

**Reply by vitorferraz10** [reply]
did you manage to solve?

**Comment by nbouvrette** [comment]
I had this error when doing a `router.replace` and managed to solve it by adding the `{ shallow: true }` option:

```ts
  useEffect(() => {
    if (asPath.endsWith('?error')) {
      setIsShowingError(true)
      void router.replace(asPath.replace('?error', ''), undefined, { shallow: true })
    }
  }, [asPath, router])
```

**Comment by zaid-ilyas64** [comment]
Facing this issue while try to get all anchor tags of the current page for external link confirmation pop up.

![image](https://user-images.githubusercontent.com/114743877/209316799-58eb65a9-9863-4489-bc06-9de55858d768.png)

**Note:** without **MutationObserver** it works fine but then some anchor tags go undetected. And if I set shallow to true routing doesn't work.

**Comment by harry-atk** [comment]
Sharing what I've observed.

When using getServerSideProps with search param routes, I'm seeing it any time a client side request for server side props is still pending (which should happen only with non-shallow routing) and another route is called. This applies to any routing on client, so if you request a non-shallow route and immediately shallow route, even to the same URL, you'll see this error because the requested page properties from the first route will not be reflected in the current page.

My use case uses shallow routing to save some page state to the URL. Without a sufficient debounce on writing to the URL, even with only shallow routing on a specific page, you can still come across this error. When you shallow route for the first time and use the browser back button to go to the first loaded URL, next seems to call for server side props on the client. I would think shallow routing and going back would still be shallow, but that doesn't seem to be the case (Might be because I'm seeing routeChangeError event before routeChangeStart is called). So any additional shallow routing, i.e. interacting with some fields that are reflected in the URL before that request finishes will cancel the request and throw the error.

**Comment by leoselig** [comment]
tl;dr Wondering if anyone also saw empty `pageProps: {}` from `getServerSideProps` in conjunction with this error

I hit this console error when trying to debug a race condition where clicking a `<Link>` while page data is still loading causes `pageProps` to be `{}` despite `getServerSideProps` clearly returning data (in our case the fields `_nextI18Next` and `dehydratedState` from react-query). I went deep into the Next source but couldn't understand where it gets lost.

**Reply by periode** [reply]
I'm having a similar issue when linking to a page with query params, and fetching data from an API through `async getServerSideProps()`

data is fetched correctly from the API, but then the props passed to the actual NextPage is `undefined`.

(fun fact: only happens in deployment, and not locally)

**Reply by leoselig** [reply]
@periode "locally" includes 'next build && next start', right? Where are you deploying to? (Amplify here)

Does it happen when navigating by clicking a link? Could you try resting the cursor on the link for a bit before clicking? Because I think I narrowed it down to a race condition concerning the link's prefetching.

**Reply by periode** [reply]
@leoselig yep, locally works like a charm. I'm deploying to a Digital Ocean app from a dockerfile

And it also happens after hovering on the link for 4-5 seconds. Idk if the fact that you're getting `{}` and i'm getting `undefined` means that we're having separate issues? Here's my relevant `getServerSideProps`:

```tsx
export async function getServerSideProps(context: GetServerSidePropsContext) {
  const syllId = context.query.sid
  const apiUrl = process.env.NEXT_PUBLIC_API_URL;
  const t = await getToken({ req: context.req, secret: process.env.NEXTAUTH_SECRET })

  const token = t ? (t.user as { _id: string, token: string }).token : '';
  const user_uuid = t ? (t.user as { _id: string, token: string })._id : '';
  const url = new URL(`syllabi/${syllId}`, apiUrl);

  const h = new Headers();
  if (t)
    h.append("Authorization", `Bearer ${token}`);

  const res = await fetch(url, { headers: h });
  if (res.ok) {
    const info = await res.json();

    if (info.user_uuid !== user_uuid)
      return { props: {} }

    return {
      props: {
        syllabusInfo: info
      }
    };
  } else {
    return {
      props: {},
    };
  }

};
```

**Reply by leoselig** [reply]
@periode Did you ever solve this? Or discover the reason this happens?

**Reply by periode** [reply]
@leoselig in the end i just cast an empty object to the required type if API request failed, seems to do the trick:

```
  const syll_res = await fetch(syllUrl, { headers: h });
  if (syll_res.ok) {
    const data = await syll_res.json();
    syllInfo = data as ISyllabus;
  } else {
    syllInfo = {} as ISyllabus
  }
```

**Comment by Yummy-sk** [comment]
I used shallow route and it solved.

**Reply by leoselig** [reply]
While true, it's not an option in all cases. Often, you want to route non-shallow. For example, (our case right now), when a page's language switcher is used. The new page props need to actually be requested for the new language's translations to be fetched.

We ended up with this... well... disgrace of a snippet in `_app.tsx`
```typescript
useEffect(() => {
    if (Object.keys(pageProps).length === 0) {
      Router.reload();
    }
  }, [pageProps]);
  ```

**Reply by nguyenbavinh-decathlon** [reply]
Hi @leoselig , thank you for your sugesstion, but I wonder how you deal with a page that doesn't have any properties added to it? In that case the web page will reload indefinitely?

**Reply by leoselig** [reply]
> Hi @leoselig , thank you for your sugesstion, but I wonder how you deal with a page that doesn't have any properties added to it? In that case the web page will reload indefinitely?

It was merely a workaround, not an actual solution. In my case, all pages had some props (e.g. i18n store). Theoretically, this workaround could have been moved to the individual pages as well.

**Comment by 521xueweihan** [comment]
This problem has troubled me for a long time. Thanks to everyone's discussion, I have an idea. I think the main problem is that `router.push` should not be executed when the server has not completed its work, but in some cases (automatically selecting the default language), it has to be done, so the method I use is to select the default language on the server and perform the international pathname address jump operation, so I use the custom server startup method and implement the specific logic in `server.js`. The code is as follows for your reference:

```typescript
/* eslint-disable @typescript-eslint/no-var-requires */
const express = require('express');
const next = require('next');
const cookieParser = require('cookie-parser');

const app = next({ dev: false });
const handle = app.getRequestHandler();

const PORT = process.env.PORT || 3000;
const DEFAULT_LOCALE = 'zh';
const FALLBACK_LOCALE = 'en';
const SUPPORTED_LOCALES = ['zh', 'en'];

async function startServer() {
  try {
    await app.prepare();
    const server = express();
    server.use(cookieParser());

    // 中间件处理语言选择和路由
    server.use((req, res, next) => {
      if (req.url.startsWith('/_next') || req.url.startsWith('/static')){
        next() 
      } else {        
        let locale = req.cookies.locale;
  
        // 解析 URL 中的语言设置
        const urlParts = req.path.split('/');
        const urlLocale = SUPPORTED_LOCALES.includes(urlParts[1]) ? urlParts[1] : null;
  
        if (!locale) {
          // 如果没有 cookie，使用 URL 中的语言或浏览器偏好
          const browserLocale = req.acceptsLanguages(SUPPORTED_LOCALES);
          locale = urlLocale || browserLocale || FALLBACK_LOCALE;
          
          // 如果浏览器语言不是中文，默认使用英语
          if (browserLocale && browserLocale !== 'zh') {
            locale = FALLBACK_LOCALE;
          }
        }
  
        // 确保 locale 是支持的语言之一
        locale = SUPPORTED_LOCALES.includes(locale) ? locale : FALLBACK_LOCALE;
        if (urlLocale) {
          // URL 中有语言前缀
          if (urlLocale !== locale) {
            // URL 语言与 cookie 不匹配，更新 locale
            locale = urlLocale;
          }
          if (locale === DEFAULT_LOCALE) {
            // 默认语言不应该在 URL 中显示，重定向到无前缀的 URL
            const newUrl = req.url.replace(`/${DEFAULT_LOCALE}`, '') || '/';
            return res.redirect(307, newUrl);
          }
        } else if (locale !== DEFAULT_LOCALE) {
          // URL 中没有语言前缀，且不是默认语言，添加前缀并重定向
          const newUrl = `/${locale}${req.url}`;
          return res.redirect(307, newUrl);
        }
  
        // 设置或更新 cookie
        res.cookie('locale', locale, { 
          maxAge: 365 * 24 * 60 * 60 * 1000,
        });
  
        req.locale = locale;
  
        // 处理尾部斜杠
        if (req.url.length > 1 && req.url.endsWith('/')) {
          req.url = req.url.slice(0, -1);
        }
        
        next();
      }

    });

    // 处理所有其他请求
    server.all('*', (req, res) => {
      return handle(req, res);
    });

    server.listen(PORT, (err) => {
      if (err) throw err;
      console.log(`> Ready on http://localhost:${PORT}`);
    });
  } catch (error) {
    console.error('Error starting server:', error);
    process.exit(1);
  }
}

startServer();
```
